src/rdwin32.h

```diff
@@ -223,37 +223,113 @@ static RD_UNUSED int rd_fd_set_nonblocking (int fd) {
-                        CloseHandle(h[0]);
-                        CloseHandle(h[1]);
-                        return (int)GetLastError();
-                }
-
-                /* Open file descriptor for handle */
-                fds[i] = _open_osfhandle((intptr_t)h[i],
-                                         i == 0 ?
-                                         O_RDONLY | O_BINARY :
-                                         O_WRONLY | O_BINARY);
-
-                if (fds[i] == -1) {
-                        CloseHandle(h[0]);
-                        CloseHandle(h[1]);
-                        return (int)GetLastError();
-                }
+        /* On windows, the "pipe" will be a tcp connection.
+        * This is to allow WSAPoll to be used to poll pipe events */
+
+        SOCKET listen_s = INVALID_SOCKET;
+        SOCKET accept_s = INVALID_SOCKET;
+        SOCKET connect_s = INVALID_SOCKET;
+
+        WSAPOLLFD poll_fd;
+
+        struct sockaddr_in listen_addr;
+        struct sockaddr_in connect_addr;
+        socklen_t sock_len = 0;
+
+        int poll_timeout_ms = 1000;
+
+        /* Create listen socket */
+        listen_s = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+        if (listen_s == INVALID_SOCKET)
+                goto err;
+
+        if (rd_fd_set_nonblocking(listen_s) != 0)
+                goto err;
+
+        listen_addr.sin_family = AF_INET;
+        listen_addr.sin_addr.s_addr = ntohl(INADDR_ANY);
+        listen_addr.sin_port = 0;
+        if (bind(listen_s, (struct sockaddr*)&listen_addr, sizeof(struct sockaddr_in)) != 0)
+                goto err;
+
+        sock_len = sizeof(struct sockaddr_in);
+        if (getsockname(listen_s, (struct sockaddr*)&connect_addr, &sock_len) != 0)
+                goto err;
+
+        if (listen(listen_s, 1) != 0)
+                goto err;
+
+        /* Create connection socket */
+        connect_s = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+        if (connect_s == INVALID_SOCKET)
+                goto err;
+
+        if (rd_fd_set_nonblocking(connect_s) != 0)
+                goto err;
+
+        connect_addr.sin_addr.s_addr = ntohl(INADDR_LOOPBACK);
+        if (connect(
+                connect_s,
+                (struct sockaddr*)&connect_addr,
+                sizeof(struct sockaddr_in)) == SOCKET_ERROR) {
+                if (WSAGetLastError() != WSAEWOULDBLOCK)
+                        goto err;
         }
+
+        /* Wait for incoming connection */
+        memset(&poll_fd, 0, sizeof(WSAPOLLFD));
+        poll_fd.fd = listen_s;
+        poll_fd.events = POLLIN;
+        if (WSAPoll(&poll_fd, 1, poll_timeout_ms) == SOCKET_ERROR)
+                goto err;
+
+        if ((poll_fd.revents & (POLLERR | POLLHUP)) > 0)
+                goto err;
+
+        accept_s = accept(listen_s, 0, 0);
+        if (accept_s == SOCKET_ERROR)
+                goto err;
+
+        /* Wait for both sockets to be writable */
```

Comment:
Why is this needed?

Reply:
In a previous version of this change I was not waiting for the connection socket to be writable, which caused an intermittent failure on rd_write immediately after creating the connection.

The connecting socket called connect as non-blocking. Event though it is connecting on loopback and we know it has been accepted, the socket must be polled to see that it is writable before the connection can be declared successful.

The accepted socket poll is not necessary and I will remove that.